File: python/두 큐합 같게만들기.py
## list를 디큐로 변경해서 빨라졌으나 변경된 시간복잡도 계산과
## 목표값을 target으로 잡아 한쪽에서만 빼고 변경하는 부분
## 왜 mx_len * 2 + 2 보다 클때인지
from collections import deque
[1,1,10,1,] 1, 
[2,2,1,2,], 3 +3 

[]

def solution(queue1, queue2):
    answer = 0
    tot = 0
    mx = 0
    a_tot = 0
    b_tot = 0 
    mx_len = max(len(queue1), len(queue2))
    
    for a in queue1:
        tot += a
        a_tot += a
        if mx < a:
            mx = a
    for b in queue2:
        tot += b
        b_tot += b
        if mx < b:
            mx = b
            
    div_target = tot/2
    if mx > div_target or div_target %1 !=0:
        answer = -1
        return answer
    
    queue1 = deque(queue1)
    queue2 = deque(queue2)
    
    while a_tot != b_tot:
        if a_tot > b_tot:
            target = queue1.popleft()
            queue2.append(target)
            a_tot -= target
            b_tot += target

        elif b_tot > a_tot:
            target = queue2.popleft()
            queue1.append(target)
            a_tot += target
            b_tot -= target

        elif a_tot == b_tot:
            return answer
        answer+=1
        if answer > mx_len * 2 + 2: #아직 이해 못한부분
            return -1
    
    return answer

File: java/두큐합같게 만들기.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = 0;
        Queue<Integer> q1 = new LinkedList<Integer>();
        Queue<Integer> q2 = new LinkedList<>();
        int size1 = queue1.length;
        int size2 = queue2.length;
        long q1Sum = 0;
        long q2Sum = 0;
        
        // System.out.println(Math.max(size1, size2) *2 +2);
        for (int i=0; i<size1; i++){
            q1.add(queue1[i]);
            q1Sum += queue1[i];
        }
        
        for (int i=0; i<size2; i++){
            q2.add(queue2[i]);
            q2Sum += queue2[i];
        }
        
    
        while(q1.size()!=0 && q2.size()!=0){
            // System.out.println(q1Sum +" "+ q2Sum);
            if(q1Sum == q2Sum){
                return answer;
            }
            if(q1Sum < q2Sum){
                int target = q2.poll();
                q1.add(target);
                q2Sum -= target;
                q1Sum += target;
            }else if(q2Sum < q1Sum){
                int target = q1.poll();
                q2.add(target);
                q1Sum -= target;
                q2Sum += target;
            }
            if(answer > ((size1 + size2) * 2 + 2) ){
                return -1;
            }
            answer+=1;
        }
        return -1;
    }
}